File: app/src/main/java/com/major/qr/adapters/DocDisplayAdapter.java
package com.major.qr.adapters;

import android.content.Context;
import android.content.res.ColorStateList;
import android.os.Handler;
import android.os.Looper;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;

import androidx.annotation.NonNull;
import androidx.core.content.ContextCompat;
import androidx.fragment.app.FragmentActivity;
import androidx.lifecycle.LifecycleOwner;
import androidx.recyclerview.widget.RecyclerView;

import com.bumptech.glide.Glide;
import com.bumptech.glide.load.engine.DiskCacheStrategy;
import com.major.qr.R;
import com.major.qr.databinding.FragmentDocumentBinding;
import com.major.qr.dialog.DocumentDetailDialog;
import com.major.qr.models.Doc;
import com.major.qr.viewmodels.DocumentViewModel;

import java.util.HashSet;
import java.util.List;

public class DocDisplayAdapter extends RecyclerView.Adapter<DocDisplayAdapter.ItemViewHolder> {
    public final String TAG = DocDisplayAdapter.class.getSimpleName();
    public List<Doc> list;
    public Context context;
    public DocumentViewModel documentViewModel;

    public HashSet<String> selectedDocs;
    FragmentDocumentBinding binding;

    public DocDisplayAdapter(Context context, List<Doc> list, DocumentViewModel documentViewModel,
                             FragmentDocumentBinding binding) {
        this.context = context;
        this.list = list;
        this.documentViewModel = documentViewModel;
        this.binding = binding;
        selectedDocs = new HashSet<>();
    }

    @NonNull
    @Override
    public ItemViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        LayoutInflater inflater = LayoutInflater.from(parent.getContext());
        View view = inflater.inflate(R.layout.item_doc, parent, false);
        return new ItemViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull ItemViewHolder holder, int position) {
        Doc doc = list.get(position);
        documentViewModel.getDocLink(doc.getDocumentReference()).observe((LifecycleOwner) context, link -> {
            doc.setDocLink(link.substring(1, link.length() - 1));
            Glide.with(context)
                    .load(doc.getDocLink())
                    .diskCacheStrategy(DiskCacheStrategy.AUTOMATIC)
                    .placeholder(R.drawable.placeholder_doc)
                    .into(holder.imageView);
        });

        holder.itemView.setOnLongClickListener(view -> {
            if (selectedDocs.contains(list.get(position).getDocumentId())) {
                unSelect(holder, position);
                return false;
            }
            select(holder, position);
            return true;
        });

        holder.itemView.setOnClickListener(view -> {
            if (selectedDocs.size() == 0) {
                DocumentDetailDialog dialog = new DocumentDetailDialog(documentViewModel, doc, this, position);
                dialog.show(((FragmentActivity) context).getSupportFragmentManager(), "Doc Dialog");
                return;
            }
            if (selectedDocs.contains(list.get(position).getDocumentId()))
                unSelect(holder, position);
            else
                select(holder, position);
        });

//        holder.updateButton.setOnClickListener(view -> {
//            UploadDialog uploadDialog = new UploadDialog(doc.getDocumentReference(), documentViewModel);
//            uploadDialog.show(((AppCompatActivity) context).getSupportFragmentManager(), "Your Dialog");
//        });


    }

    @Override
    public int getItemCount() {
        return list.size();
    }

    public void select(ItemViewHolder holder, int position) {
//        holder.itemView.setBackgroundColor(ContextCompat.getColor(context, R.color.teal_200));
        holder.itemView.setBackgroundTintList(ColorStateList.valueOf(
                ContextCompat.getColor(context, R.color.teal_200)));
        selectedDocs.add(list.get(position).getDocumentId());
        binding.clearAll.setVisibility(View.VISIBLE);
        new Handler(Looper.getMainLooper()).postDelayed(() -> {
            holder.itemView.setBackgroundResource(R.drawable.attendance_background);
        }, 400);
    }

    public void unSelect(ItemViewHolder holder, int position) {
        holder.itemView.setBackgroundTintList(ColorStateList.valueOf(
                ContextCompat.getColor(context, R.color.primary_blue)));
        selectedDocs.remove(list.get(position).getDocumentId());
        if (selectedDocs.size() == 0)
            binding.clearAll.setVisibility(View.GONE);
//        holder.itemView.setBackgroundResource(R.drawable.document_background);
        new Handler(Looper.getMainLooper()).postDelayed(() -> {
            holder.itemView.setBackgroundResource(R.drawable.document_background);
        }, 400);
    }

    public static class ItemViewHolder extends RecyclerView.ViewHolder {
        ImageView imageView;

        public ItemViewHolder(View itemView) {
            super(itemView);
            imageView = itemView.findViewById(R.id.imageView);
        }
    }
}

File: app/src/main/java/com/major/qr/adapters/DocDisplayAdapter.kt
package com.major.qr.adapters

import android.content.Context
import android.content.res.ColorStateList
import android.os.Handler
import android.os.Looper
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import androidx.appcompat.content.res.AppCompatResources
import androidx.core.content.ContextCompat
import androidx.fragment.app.FragmentActivity
import androidx.lifecycle.LifecycleOwner
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.bumptech.glide.load.engine.DiskCacheStrategy
import com.major.qr.R
import com.major.qr.databinding.FragmentDocumentBinding
import com.major.qr.dialog.DocumentDetailDialog
import com.major.qr.models.Doc
import com.major.qr.viewmodels.DocumentViewModel

class DocDisplayAdapter(
    var context: Context, var list: ArrayList<Doc>, var documentViewModel: DocumentViewModel,
    var binding: FragmentDocumentBinding
) : RecyclerView.Adapter<DocDisplayAdapter.ItemViewHolder>() {
    val TAG = DocDisplayAdapter::class.java.simpleName
    @JvmField
    var selectedDocs: HashSet<String?>

    init {
        selectedDocs = HashSet()
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ItemViewHolder {
        val inflater = LayoutInflater.from(parent.context)
        val view = inflater.inflate(R.layout.item_doc, parent, false)
        return ItemViewHolder(view)
    }

    override fun onBindViewHolder(holder: ItemViewHolder, position: Int) {
        val doc = list[position]
        if (doc.documentType == "pdf") holder.imageView.setImageDrawable(
            AppCompatResources.getDrawable(context, R.drawable.pdf_placeholder)
        ) else documentViewModel.getDocLink(doc.documentReference!!).observe(
            (context as LifecycleOwner)
        ) { link: String ->
            doc.docLink = link.substring(1, link.length - 1)
            Glide.with(context)
                .load(doc.docLink)
                .diskCacheStrategy(DiskCacheStrategy.AUTOMATIC)
                .placeholder(R.drawable.placeholder_doc)
                .into(holder.imageView)
        }
        holder.itemView.setOnLongClickListener { view: View? ->
            if (selectedDocs.contains(list[position].documentId)) {
                unSelect(holder, position)
                return@setOnLongClickListener false
            }
            select(holder, position)
            true
        }
        holder.itemView.setOnClickListener { view: View? ->
            if (selectedDocs.size == 0) {
                val dialog = DocumentDetailDialog(
                    documentViewModel, doc,
                    this, position
                )
                dialog.show((context as FragmentActivity).supportFragmentManager, "Doc Dialog")
                return@setOnClickListener
            }
            if (selectedDocs.contains(list[position].documentId)) unSelect(
                holder,
                position
            ) else select(holder, position)
        }
    }

    override fun getItemCount(): Int {
        return list.size
    }

    fun select(holder: ItemViewHolder, position: Int) {
//        holder.itemView.setBackgroundColor(ContextCompat.getColor(context, R.color.teal_200));
        holder.itemView.backgroundTintList = ColorStateList.valueOf(
            ContextCompat.getColor(context, R.color.teal_200)
        )
        selectedDocs.add(list[position].documentId)
        binding.clearAll.visibility = View.VISIBLE
        Handler(Looper.getMainLooper()).postDelayed(
            { holder.itemView.setBackgroundResource(R.drawable.attendance_background) },
            400
        )
    }

    fun unSelect(holder: ItemViewHolder, position: Int) {
        holder.itemView.backgroundTintList = ColorStateList.valueOf(
            ContextCompat.getColor(context, R.color.primary_blue)
        )
        selectedDocs.remove(list[position].documentId)
        if (selectedDocs.size == 0) binding.clearAll.visibility = View.GONE
        Handler(Looper.getMainLooper()).postDelayed(
            { holder.itemView.setBackgroundResource(R.drawable.document_background) },
            400
        )
    }

    class ItemViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        var imageView: ImageView

        init {
            imageView = itemView.findViewById(R.id.imageView)
        }
    }
}